// @shopify/react-graphql/dist/src/links.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var apollo_link_1 = require("apollo-link");
function createSsrExtractableLink() {
    return new SsrExtractableLink();
}
exports.createSsrExtractableLink = createSsrExtractableLink;
var SsrExtractableLink = /** @class */ (function (_super) {
    tslib_1.__extends(SsrExtractableLink, _super);
    function SsrExtractableLink() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.operations = new Set();
        return _this;
    }
    SsrExtractableLink.prototype.resolveAll = function (then) {
        if (this.operations.size > 0) {
            return Promise.all(tslib_1.__spread(this.operations)).then(then);
        }
        return then();
    };
    SsrExtractableLink.prototype.request = function (operation, nextLink) {
        var _this = this;
        if (nextLink == null) {
            throw new Error('SsrExtractableLink must not be the last link in the chain.');
        }
        var operationDone;
        var promise = new Promise(function (resolve) {
            operationDone = function () {
                _this.operations.delete(promise);
                resolve();
            };
        });
        this.operations.add(promise);
        return new apollo_link_1.Observable(function (observer) {
            return nextLink(operation).subscribe({
                complete: function () {
                    observer.complete();
                    operationDone();
                },
                next: observer.next.bind(observer),
                error: observer.next.bind(observer),
            });
        });
    };
    return SsrExtractableLink;
}(apollo_link_1.ApolloLink));
exports.SsrExtractableLink = SsrExtractableLink;
