// @shopify/react-testing/build/cjs/matchers/props.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.toHaveReactProps = toHaveReactProps;
exports.toHaveReactDataProps = toHaveReactDataProps;

var _jestMatcherUtils = require("jest-matcher-utils");

var _utilities = require("./utilities");

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function toHaveReactProps(node, props) {
  var _this = this;

  (0, _utilities.assertIsNode)(node, {
    expectation: 'toHaveReactProps',
    isNot: this.isNot
  });

  if (props == null || _typeof(props) !== 'object') {
    return {
      pass: false,
      message: function message() {
        return "You passed ".concat(props == null ? String(props) : "a ".concat(_typeof(props)), " as props, but it must be an object.");
      }
    };
  }

  var pass = Object.keys(props).every(function (key) {
    return _this.equals(props[key], node.props[key]);
  });
  var message = pass ? function () {
    return "".concat((0, _jestMatcherUtils.matcherHint)('.not.toHaveReactProps', node.toString()), "\n\n") + "Expected the React element:\n  ".concat((0, _jestMatcherUtils.RECEIVED_COLOR)(node.toString()), "\n") + "Not to have props:\n  ".concat((0, _jestMatcherUtils.printExpected)(props), "\n") + "Received:\n  ".concat((0, _jestMatcherUtils.printReceived)(node.props), "\n");
  } : function () {
    var diffString = (0, _utilities.diffPropsForNode)(node, props, {
      expand: _this.expand
    });
    return "".concat((0, _jestMatcherUtils.matcherHint)('.toHaveReactProps', node.toString()), "\n\n") + "Expected the React element:\n  ".concat((0, _jestMatcherUtils.RECEIVED_COLOR)(node.toString()), "\n") + "To have props:\n  ".concat((0, _jestMatcherUtils.printExpected)(props), "\n") + "Received:\n  ".concat((0, _jestMatcherUtils.printReceived)(node.props), "\n").concat(diffString ? "Difference:\n".concat(diffString, "\n") : '');
  };
  return {
    pass: pass,
    message: message
  };
}

function toHaveReactDataProps(node, data) {
  return toHaveReactProps.call(this, node, data);
}
