// @shopify/react-testing/build/node/matchers/utilities.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.assertIsNode = assertIsNode;
exports.assertIsType = assertIsType;
exports.diffs = diffs;
exports.printType = printType;
exports.diffPropsForNode = diffPropsForNode;
exports.pluralize = pluralize;

var _jestMatcherUtils = require("jest-matcher-utils");

var _root = require("../root");

var _element = require("../element");

function assertIsNode(node, {
  expectation,
  isNot
}) {
  if (node == null) {
    throw new Error((0, _jestMatcherUtils.matcherErrorMessage)((0, _jestMatcherUtils.matcherHint)(`.${expectation}`, undefined, undefined, {
      isNot
    }), `${(0, _jestMatcherUtils.RECEIVED_COLOR)('received')} value must be an @shopify/react-testing Root or Element object`, `Received ${(0, _jestMatcherUtils.RECEIVED_COLOR)('null')}.\nThis usually means that your \`.findX\` method failed to find any matching elements.`));
  }

  if (Array.isArray(node) && node.length > 1 && (node[0] instanceof _root.Root || node[0] instanceof _element.Element)) {
    throw new Error((0, _jestMatcherUtils.matcherErrorMessage)((0, _jestMatcherUtils.matcherHint)(`.${expectation}`, undefined, undefined, {
      isNot
    }), `${(0, _jestMatcherUtils.RECEIVED_COLOR)('received')} value must be an @shopify/react-testing Root or Element object`, `Received an ${(0, _jestMatcherUtils.RECEIVED_COLOR)('array of Root or Element objects')}.\nThis usually means that you passed in the result of \`.findAllX\`. Pass the result of \`.findX\` instead.`));
  }

  if (!(node instanceof _root.Root) && !(node instanceof _element.Element)) {
    throw new Error((0, _jestMatcherUtils.matcherErrorMessage)((0, _jestMatcherUtils.matcherHint)(`.${expectation}`, undefined, undefined, {
      isNot
    }), `${(0, _jestMatcherUtils.RECEIVED_COLOR)('received')} value must be an @shopify/react-testing Root or Element object`, (0, _jestMatcherUtils.printWithType)('Received', node, _jestMatcherUtils.printReceived)));
  }
}

function assertIsType(type, {
  expectation,
  isNot
}) {
  if (type == null) {
    throw new Error((0, _jestMatcherUtils.matcherErrorMessage)((0, _jestMatcherUtils.matcherHint)(`.${expectation}`, undefined, undefined, {
      isNot
    }), `${(0, _jestMatcherUtils.RECEIVED_COLOR)('expected')} value must be a string or a valid React component.`, (0, _jestMatcherUtils.printWithType)('Expected', type, _jestMatcherUtils.printExpected)));
  }
}

function diffs(element, props, expand) {
  return element.reduce((diffs, element, index) => {
    const separator = index === 0 ? '' : '\n\n';
    return `${diffs}${separator}${normalizedDiff(element, props, {
      expand,
      showLegend: index === 0
    })}`;
  }, '');
}

function normalizedDiff(element, props, {
  expand = false,
  showLegend = false
}) {
  const result = diffPropsForNode(element, props, {
    expand
  }) || '';
  return showLegend ? result : result.split('\n\n')[1];
}

function printType(type) {
  if (typeof type === 'object' && '_context' in type) {
    const context = type._context;
    const componentName = type === context.Provider ? 'Provider' : 'Consumer';
    const displayName = context.displayName || 'Context';
    return `<${displayName}.${componentName} />`;
  }

  const displayName = typeof type === 'string' ? type : type.displayName || type.name || 'Component';
  return `<${displayName} />`;
}

function diffPropsForNode(node, props, {
  expand = false
}) {
  return (0, _jestMatcherUtils.diff)(props, getObjectSubset(node.props, props), {
    expand
  });
} // Original from https://github.com/facebook/jest/blob/master/packages/expect/src/utils.ts#L107


function getObjectSubset(object, subset) {
  if (Array.isArray(object)) {
    if (Array.isArray(subset) && subset.length === object.length) {
      return subset.map((sub, i) => getObjectSubset(object[i], sub));
    }
  } else if (object instanceof Date) {
    return object;
  } else if (typeof object === 'object' && object !== null && typeof subset === 'object' && subset !== null) {
    const trimmed = {};
    Object.keys(subset).filter(key => Reflect.has(object, key)).forEach(key => trimmed[key] = getObjectSubset(object[key], subset[key]));

    if (Object.keys(trimmed).length > 0) {
      return trimmed;
    }
  }

  return object;
}

function pluralize(word, count) {
  return count === 1 ? word : `${word}s`;
}
